python/cudf/cudf/core/series.py

```diff
@@ -3275,7 +3275,7 @@ def value_counts(
         res.index.name = self.name
 
         if sort:
-            res = res.sort_values(ascending=ascending)
+            res = res.sort_index().sort_values(ascending=ascending)
```

Comment:
It seems like you're trying to enforce stable index ordering, and that index ordering is always ascending. Is that correct?

`value_counts` is a pretty popular API so I want to ensure we keep good performance here. Rather than calling sort twice, can we call sort one time where we include the index as the least-important column key in the sort, and then reconstruct the result with that index column as the index? (Set ascending correctly for that index column.) Maybe there are other places in the Python code where this is being done?

Reply:
@bdice Thanks for catching this! You’re right about the performance concern. `.sort_index()` was only added to make the docstring test deterministic, my hash changes caused bins with tied counts in `value_counts(bins=3)` to appear in a different order.

Should cudf `value_counts()` guarantee ordering for tied counts? If not, we can remove `.sort_index()` and sort only in the docstring test. Otherwise, we could do a single sort using the index as a secondary key for stable ordering as you suggested.

What do you think is the best approach?

Reply:
Here's some helpful info. Looks like pandas preserves the input order for values of equal counts.

```python
>>> pd.Series([3, 3, 4, 4]).value_counts(sort=True)
3    2
4    2
Name: count, dtype: int64
>>> pd.Series([4, 4, 3, 3]).value_counts(sort=True)
4    2
3    2
Name: count, dtype: int64
```

Reply:
And just to be clear, this isn't a performance concern but a correctness concern (if we care about input order preservation).

`res.sort_index().sort_values(ascending=ascending)` won't correctly preserve input order.

Reply:
Our rule of thumb these days is that this sort of behavior should be gated behind pandas compatibility mode, which is automatically turned on when running with cudf.pandas. So hiding this sorting behind a conditional like [this one](https://github.com/rapidsai/cudf/blob/e4b984e9ad24b03bcb73ebfbe8b94b0354bf2665/python/cudf/cudf/core/dataframe.py#L1263) would be appropriate.